app/controllers/uploads_controller.rb

```diff
@@ -187,6 +192,81 @@ def metadata
     }
   end
 
+  def generate_presigned_put
+    return render_404 if !SiteSetting.enable_direct_s3_uploads
+
+    file_name = params.require(:file_name)
+    type = params.require(:type)
+
+    RateLimiter.new(
+      current_user, "generate-presigned-put-upload-stub", PRESIGNED_PUT_RATE_LIMIT_PER_MINUTE, 1.minute
+    ).performed!
+
+    # don't want people posting arbitrary S3 metadata so we just take the
+    # one we need. all of these will be converted to x-amz-meta- metadata
+    # fields in S3 so it's best to use dashes in the names for consistency
+    #
+    # this metadata is baked into the presigned url and is not altered when
+    # sending the PUT from the clientside to the presigned url
+    metadata = if params[:metadata].present?
+      meta = {}
+      if params[:metadata]["sha1-checksum"].present?
+        meta["sha1-checksum"] = params[:metadata]["sha1-checksum"]
+      end
+      meta
+    end
+
+    url = Discourse.store.signed_url_for_temporary_upload(
+      file_name, metadata: metadata
+    )
+    key = Discourse.store.path_from_url(url)
+
+    upload_stub = ExternalUploadStub.create!(
+      key: key,
+      created_by: current_user,
+      original_filename: file_name,
+      upload_type: type
+    )
+
+    render json: { method: :put, url: url, key: key, unique_identifier: upload_stub.unique_identifier }
+  end
+
+  def complete_external_upload
+    return render_404 if !SiteSetting.enable_direct_s3_uploads
+
+    unique_identifier = params.require(:unique_identifier)
+
+    external_upload_stub = ExternalUploadStub.find_by(unique_identifier: unique_identifier)
+    return render_404 if external_upload_stub.blank?
+
+    raise Discourse::InvalidAccess if external_upload_stub.created_by_id != current_user.id
+    external_upload_manager = ExternalUploadManager.new(external_upload_stub)
+
+    hijack do
+      begin
+        upload = external_upload_manager.promote_to_upload!
+        if upload.errors.empty?
+          external_upload_manager.destroy!
+          render json: UploadsController.serialize_upload(upload), status: 200
+        else
+          render_json_error(upload.errors.to_hash.values.flatten, status: 422)
+        end
+      rescue ExternalUploadManager::ChecksumMismatchError => err
+        Discourse.warn_exception(err, message: I18n.t("upload.checksum_mismatch_failure"))
```

Comment:
Do the errors here indicate a bug in an application that we can fix given the information which we're going to be logging here? If not, I worry about warning in the logs because it'll just be noise which we can't do anything about.

Reply:
Hmm you are right, can't really do anything about this. I had these errors returning to the user originally but Robin preferred we didn't. Will just change this to a debug log level maybe?

Reply:
I think if the logs are not useful to us, we shouldn't log by default as it'll just add noise to the logs. Maybe add the logging behind a hidden site setting if we ever need to debug? If not I suggest we just remove it completely until a use case shows up.

Reply:
Okay hid them behind a site setting